differs/rpm_diff.go

```diff
@@ -87,7 +95,65 @@ func (a RPMAnalyzer) getPackages(image pkgutil.Image) (map[string]util.PackageIn
 		}
 	}
 
-	return rpmDataFromContainer(image)
+	packages, err := rpmDataFromImageFS(image)
+	if err != nil {
+		logrus.Warn("Trying to run the RPM binary of the image in a container")
+		return rpmDataFromContainer(image)
+	}
+	return packages, err
+}
+
+// rpmDataFromImageFS runs a local rpm binary, if any, to query the image
+// rpmdb and returns a map of installed packages.
+func rpmDataFromImageFS(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	packages := make(map[string]util.PackageInfo)
+	// Check there is an executable rpm tool in host
+	if err := exec.Command("rpm", "--version").Run(); err != nil {
+		logrus.Warn("No RPM binary in host")
+		return packages, err
+	}
+	dbPath, err := rpmDBPath(image.FSPath)
+	if err != nil {
+		logrus.Warnf("Couldn't find RPM database: %s", err.Error())
+		return packages, err
+	}
+	cmdArgs := append([]string{"--root", image.FSPath, "--dbpath", dbPath}, rpmCmd[1:]...)
+	out, err := exec.Command(rpmCmd[0], cmdArgs...).Output()
+	if err != nil {
+		logrus.Warnf("RPM call failed: %s", err.Error())
+		return packages, err
+	}
+	output := strings.Split(string(out), "\n")
+	return parsePackageData(output)
+}
+
+// rpmDBPath tries to get the RPM database path from the /usr/lib/rpm/macros
+// file in the image rootfs.
+func rpmDBPath(rootFSPath string) (string, error) {
+	rpmMacros, err := os.Open(filepath.Join(rootFSPath, "usr/lib/rpm/macros"))
```

Comment:
is it possible that the RPM database could have been put in a different location by the creator of the image? are there other common locations that we might want to check here as well?

Reply:
Here we are looking for the value of %_dbpath macro which sets the path of the database. 

> is it possible that the RPM database could have been put in a different location by the creator of the image?

Yes, it could be the case that the user sets its own %_dbpath in a custom user defined macro. As far as I saw in fedora, opensuse and centos, the %_dbpath defined in /usr/lib/rpm/macros is not overwritten in any of them. According to the man pages RPM looks for macros in:

> The  default FILELIST is /usr/lib/rpm/macros:/usr/lib/rpm/macros.d/macros.\*:/usr/lib/rpm/platform/%{_target}/macros:/usr/lib/rpm/fileattrs/\*.attr:/usr/lib/rpm/redhat/macros:/etc/rpm/macros.\*:/etc/rpm/macros:/etc/rpm/%{_target}/macros:~/.rpmmacros

The macros from the image could be loaded by using the `--macros` rpm flag and passing the appropriate paths list. However I encountered that this flag expects a colon separated list of paths and I could not manage to escape colons that they are part of the path we want to provide (i.e providing `--macros opensuse@sha256:8e70aa9ec2/usr/lib/rpm/macros` is understood as two separate paths :man_facepalming: and I could not manage to escape the colons).

That's the reason to manually parse the `/usr/lib/rpm/macros` file to figure out the rpmdb path. Note that then is verified the path exists and if the command fails it falls back to running the container. I tested this approach of parsing the main macros file in opensuse, centos and fedora with success.